lib/hutch/broker.rb

```diff
@@ -215,41 +216,8 @@ def nack(delivery_tag)
       channel.nack(delivery_tag, false, false)
     end
 
-    def log_publication(serializer, payload, routing_key)
-      logger.info {
-        spec =
-          if serializer.binary?
-            "#{payload.bytesize} bytes message"
-          else
-            "message '#{payload}'"
-          end
-        "publishing #{spec} to #{routing_key}"
-      }
-    end
-
-    def publish(routing_key, message, properties = {}, options = {})
-      ensure_connection!(routing_key, message)
-
-      serializer = options[:serializer] || @config[:serializer]
-
-      non_overridable_properties = {
-        routing_key:  routing_key,
-        timestamp:    connection.current_timestamp,
-        content_type: serializer.content_type,
-      }
-      properties[:message_id]   ||= generate_id
-
-      payload = serializer.encode(message)
-
-      log_publication(serializer, payload, routing_key)
-
-      response = exchange.publish(payload, {persistent: true}.
-        merge(properties).
-        merge(global_properties).
-        merge(non_overridable_properties))
-
-      channel.wait_for_confirms if @config[:force_publisher_confirms]
-      response
+    def publish(*args)
+      publisher.new(connection, channel, exchange, @config).publish(*args)
```

Comment:
@michaelklishin by calling `#declare_publisher!` just after `#declare_exchange!`, we get the behavior your asking for.

However, there are tests for publishing without a connection that break with this change. These tests work by calling `#publish` without having called `#connect`. Without calling `#connect`, `@publisher` remains nil, and so we get `undefined method #publish for nil:NilClass` rather than a Hutch::PublishError.

To repair this test, I think we'd need to modify the test so that #connect gets called but obtains an invalid connection (rather than no connection) **-or-** we'd need to ensure we wrap `Broker#publish` with `#with_connection_error_handler` and be sure to raise the appropriate error, rather than letting `NoMethodError` be raised.  IMO the memoization is much cleaner, but I'll do whatever you suggest.  What do you suggest?